### PR TITLE
treewide: gobject-introspection from buildInputs to nativeBuildInputs

### PR DESCRIPTION
**Pull request by Artturin**
gobject-introspection should be in nativeBuildInputs for cross to work properly (so propagations and hook work properly)

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/lua-modules/overrides.nix
Comment (jtojnar) [inline pkgs/development/lua-modules/overrides.nix:183]: This likely uses libgirepository as well.

File: pkgs/desktops/cinnamon/cjs/default.nix
Comment (Artturin) [reply pkgs/desktops/cinnamon/cjs/default.nix:39]: a typelib is installed

File: pkgs/development/tools/analysis/tartan/default.nix
Comment (jtojnar) [inline pkgs/development/tools/analysis/tartan/default.nix:32]: This definitely uses g-i just to link libgirepository.

File: pkgs/development/tools/poetry2nix/poetry2nix/overrides/default.nix
Comment (jtojnar) [inline pkgs/development/tools/poetry2nix/poetry2nix/overrides/default.nix:1880]: Again, libgirepository.

File: pkgs/desktops/cinnamon/cjs/default.nix
Comment (jtojnar) [reply pkgs/desktops/cinnamon/cjs/default.nix:39]: Only a private one that will not be picked up by the setup hook. And the hook is not propagated either.

Comment (jtojnar) [reply pkgs/desktops/cinnamon/cjs/default.nix:39]: Yeah, that is correct. gi probably needs to be in both here.

Comment (Artturin) [reply pkgs/desktops/cinnamon/cjs/default.nix:39]:
```diff
diff --git a/pkgs/desktops/cinnamon/cjs/default.nix b/pkgs/desktops/cinnamon/cjs/default.nix
index 681d6afd4ead..1b3aa41221b1 100644
--- a/pkgs/desktops/cinnamon/cjs/default.nix
+++ b/pkgs/desktops/cinnamon/cjs/default.nix
@@ -33,10 +33,13 @@ stdenv.mkDerivation rec {
     pkg-config
     which # for locale detection
     libxml2 # for xml-stripblanks
-    gobject-introspection
+    dbus
+    glib
   ];
 
+  strictDeps = true;
   buildInputs = [
+    gobject-introspection # for library
     cairo
     readline
     spidermonkey_102
```

doesn't fail, can't test cross because it fails `meson.build:261:22: ERROR: Can not run test applications in this cross environment.`

Comment (Artturin) [reply pkgs/desktops/cinnamon/cjs/default.nix:39]: added mesonEmulatorHook and not it fails with g-ir-scanner

Comment (Artturin) [reply pkgs/desktops/cinnamon/cjs/default.nix:39]: gobject-introspection in nativeBuildInputs already propagates itself to buildInputs, so no need to have it in both, which is why i would prefer having it in nativeBuildInputs always because then cross would always just work

Comment (jtojnar) [reply pkgs/desktops/cinnamon/cjs/default.nix:39]: I generally prefer explicit direct dependencies over propagation since it paints a clearer picture. But cjs is not my package so I will leave it up to its maintainers.

Comment (mkg20001) [reply pkgs/desktops/cinnamon/cjs/default.nix:39]: > because then cross would always just work

if having it twice breaks a particular usecase then let's just have it in nativeBuildInputs

otherwise whatever @bobby285271 prefers

**Review by SuperSandro2000 — APPROVED** [review]

**Pull request merged**
